dashboard/lib/services/lti/auth_id_generator.rb

```diff
@@ -0,0 +1,52 @@
+require 'cdo/honeybadger'
+
+module Services
+  module Lti
+    class AuthIdGenerator
```

Comment:
What's the benefit of having this as a class vs a module? It seems to act like a function that takes a JWT and returns an ID, and in the code we instantiate a fresh `AuthIdGenerator` each time we generate an ID, so would it fit better as a `generate` function within a `Services::Lti::AuthIdGenerator`?

Reply:
As per @artem-vavilov suggestion, I followed the [Service object structure pattern](https://blog.trk.in.rs/2018/11/30/object-oriented-design-principles-and-patterns/#service-objects)

Reply:
Ruby is a purely object-oriented language (literally everything in Ruby is an object). While there is nothing inherently wrong with employing a functional programming style in Ruby, especially considering its potential for greater memory efficiency and simplicity, the object-oriented approach is generally preferred.

I believe that, due to the complexity introduced by the recent update, the `.generate_auth_id` business logic now necessitates encapsulation in its own class.

I recommended employing an object-oriented approach for the `Services::Lti::AuthIdGenerator` because it centralizes all operations related to generating the authentication ID within a single, coherent class instance. This encapsulation makes the code more intuitive and manageable. Additionally, it facilitates sharing the `id_token` across various operations within the class, enhancing data consistency and accessibility.

Reply:
I definitely agree it makes sense to break the code down into manageable chunks. I'm not fully convinced for this particular class since it looks to have more boilerplate than actual code, but since this is the preferred pattern let's roll with it and see. I imagine in other cases, the benefit will be clearer.

Reply:
I recently was trying to decide between a class and a module on another feature. ChatGPT says: 
* Use modules for grouping functions, mixins, and namespaces without needing instances.
* Use classes when you need to instantiate objects, model entities with specific attributes and behaviors, or establish an inheritance hierarchy.

In our case, all we need this functionality to do is take in an id_token, and return a string. Instantiating an object (class) for this seems like overkill to me.

Reply:
All we need to do now, is:
1. Take an `id_token`.
2. Subtract the `audience` depend on its type.
3. Validate the `audience` for each type according to specific conditions.
4. And only then return a string.

In my humble opinion, implementing this logic within an instance makes the code more maintainable, understandable, and extensible for possible future changes.

I would also like to hear the opinions of @snickell and @Hamms this.

Reply:
Probably it doesn't matter either way. When in doubt, I usually go with "whoever wrote the code did it the right way". I _personally_ would have most likely written it as a single method, but not because it would be better, that's just how **I** would have happened to do it.

My personal opinion: this is totally reasonable as-is, no need for too many of us to measure the code too closely, just go with it as it is written

1. I like how clearly call reads in this instance
2. I like how clearly issuer, audience and subject are broken into named bits
3. I like that id_token doesn't need to be piped around to each method, which is part of why call is so clear
4. I don't love the idea of invoking it as `AuthIdGenerator.new.call()`, but everything is a balance, and it doesn't bother me either.

Reply:
We can invoke the service as `AuthIdGenerator.call()`: https://thiagolimadeveloper.medium.com/calling-services-object-without-instance-in-ruby-ca8b3f95fd10

Reply:
:tada: for Service Objects!

+1 to what Artem and Seth are saying; I agree it would probably work fine to implement this as a stateless module that manually passes the token around to whatever needs it, but I think this implementation does a good job of reflecting standard Ruby design principles. It might take a few more lines of code than it otherwise would, but I think those lines are being spent to improve readability rather than hinder it.